### PR TITLE
[ASCollectionView] Add Method to Wait for Updates During Next Layout Pass

### LINKED ISSUE #1584: [ASCollectionNode / ASTableNode] Method to Wait for Initial Nodes to Be Loaded
`waitUntilAllUpdatesAreCommitted` is great! Could we go even further and make a variant that waits until the initial nodes are all measured and inserted as well?

I've got a pager node with 3 tabs, and I'd like to push a VC already jumped to the third tab, but I can't. I add `waitUntilAllUpdatedAreCommitted` in `viewWillAppear` and it has no effect, because `_editingTransactionQueue` is empty but numberOfSections is 0. `viewDidLayoutSubviews` is no help either because my view controller's view is a container that holds the pager inside it. Currently my best option is to async out of `viewWillAppear:` and hope that by then the nodes are ready, else to poll repeatedly and jump as soon as I see `numberOfItemsInSection:` is correct but these aren't ideal.

@nguyenhuy Any advice?

### PR DESCRIPTION
**Pull request by Adlai-Holler**
During the normal flow, the initial data is loaded during the first layout pass. This adds a flag for the user to tell us "when the next layout pass happens, block the main thread until the data is up-to-date".

- Simplifies our ASCollectionView testing setup by combining ASCollectionViewTestController and ASCollectionViewTestDelegate into just the controller.
- Fixes an issue where `waitUntilAllUpdatesAreCommitted` fails if the initial reload hasn't been performed yet.
- Minor: makes `reloadData` only be callable on the main thread for ASCollectionView. It's a pretty small API change, since all of our other editing operations are already required to be on the main thread. It was already required for table view.

Resolves #1584 and essentially #1642 also.

Ready for review!

### CODE REVIEW

**Comment by Adlai-Holler** [comment]
@maicki @levi @appleguy thoughts? It'll be important to raise the visibility of this option because it's very important. We may even want to connect it to ASViewController's `neverShowPlaceholders` flag since this is part of that.

File: AsyncDisplayKit/ASCollectionView.mm [hunk 1]
```diff
@@ -520,6 +533,13 @@ - (void)selectItemAtIndexPath:(NSIndexPath *)indexPath animated:(BOOL)animated s
   [super selectItemAtIndexPath:indexPath animated:animated scrollPosition:scrollPosition];
 }
 
+- (void)waitForUpdatesDuringNextLayoutPass
+{
+  ASDisplayNodeAssertMainThread();
+  [self setNeedsLayout];
```
Comment (appleguy) [inline AsyncDisplayKit/ASCollectionView.mm]: @Adlai-Holler should this actually set a new layout pass as being needed?  The name doesn't currently imply that.

Comment (Adlai-Holler) [inline AsyncDisplayKit/ASCollectionView.mm]: You're absolutely right – that was a holdover from some previous implementation I believe. Removed.

**Comment by Adlai-Holler** [comment]
Ready for next review @maicki @appleguy @garrettmoon

**Comment by maicki** [comment]
LGTM

**Comment by maicki** [comment]
LGTM

File: AsyncDisplayKit/ASCollectionView.h [hunk 2]
```diff
@@ -52,6 +52,16 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, weak) id<ASCollectionDataSource> asyncDataSource;
 
 /**
+ * Tell the collection view that, during the next layoutSubviews pass, it should block the
+ * main thread and wait for all items to be up-to-date.
+ *
+ * @discussion This is useful when you want to ensure that the user never sees an empty
+ * collection view. It is better to call this than to explicitly wait using `waitUntilAllUpdatesAreCommitted`
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASCollectionView.h]: Can we add a `@warning` in the waitUntilAllUpdatesAreCommitted method indicating they might want to use this method instead?

**Comment by Adlai-Holler** [comment]
This will be changed so much in light of the 2.0 ASCollectionView/ASTableView -> Node diff that it may as well be closed. Sayonara.

**Pull request closed without merging**